src/stores/userImage.ts

```diff
@@ -0,0 +1,354 @@
+
+      // Register three observers:
+      // 1. 'state_changed' observer, called any time the state changes
+      // 2. Error observer, called on failure
+      // 3. Completion observer, called on successful completion
+      uploadTask.on(
+        'state_changed',
+        (snapshot) => {
+          // Observe state change events such as progress, pause, and resume
+          // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
+          const progress =
+            (snapshot.bytesTransferred / snapshot.totalBytes) * 100
+          console.log(`Upload is ${progress}% done`)
+          switch (snapshot.state) {
+            case 'paused':
+              console.log('Upload is paused')
+              break
+            case 'running':
+              console.log('Upload is running')
+              break
+            default:
+          }
+        },
+        (error) => {
+          // A full list of error codes is available at
+          // https://firebase.google.com/docs/storage/web/handle-errors
+          switch (error.code) {
+            case 'storage/unauthorized':
+              // User doesn't have permission to access the object
+              break
+            case 'storage/canceled':
+              // User canceled the upload
+              break
+
+            // ...
+
+            case 'storage/unknown':
+              // Unknown error occurred, inspect error.serverResponse
+              break
+            default:
+          }
+        },
+        () => {
+          // Handle successful uploads on complete
+          // For instance, get the download URL: https://firebasestorage.googleapis.com/...
+          getDownloadURL(uploadTask.snapshot.ref)
+            .then((downloadURL) => {
+              // firestoreのimageデータを更新
+              this.uploadImageToFirestore(downloadURL, file, imageId).catch(
+                (error) => {
+                  console.log(error.code)
+                },
+              )
+            })
+            .catch((error) => {
+              console.log(error.code)
+            })
+        },
+      )
+    },
+
+    // Storageで作成したurlを受け取り、firestoreに保存する
+    async uploadImageToFirestore(url: string, file: File, imageId: string) {
+      const { uid } = storeToRefs(useAuthStore())
+      const id = imageId
+      const fileExtention = file.name.split('.').pop() as string // input accept .jpeg .png .jpg
+
+      const newImage: UploadedImage = {
+        userUid: uid.value,
+        id, // imageId
+        storageURL: url, // StorageのURL
+        fileName: file.name,
+        fileType: file.type,
+        fileExtention,
+        createdAt: Timestamp.now(),
+        show: true, // Toolbarに表示・非表示
+        canvases: {}, // key: canvasId, value: 使用数 valueが0になったらkeyを削除
+      }
+      // フロント側でも追加
+      this.uploadedImages = [...this.uploadedImages, newImage]
+      try {
+        const docRef = doc(db, 'userImageStorage', uid.value)
```

Comment:
6回登場しているので、32行目 exportの前に宣言して良いと思います

Reply:
uidと同じ理由で宣言するとエラーになるので
docRefを呼び出すメソッドを作ってみましたが
なんか逆にわかりづらくなる気が…
```suggestion  
        getDocRef(uid: Ref<string>) {
          const docRef = doc(db, 'userImageStorage', uid.value)
          return docRef
        },
        ...
        try {
          const docRef = this.getDocRef(uid)
          const userImages = await this.getImageDocSnap(docRef)
        ...
```